### ISSUE TITLE
Make the controller tolerant to malformed resources

### ISSUE THREAD

**Issue by alexec** [argoproj/argo#3677]
Currently, if you `kubectl apply -f malformed-cronworkflow.yaml` (or workflow template or cluster workflow template) we cannot list those resources, see #3666.

The workaround is delete the malformed resource. However, in a multi-tenant environment, this means a single malformed resource could other user's workflows.

# Analysis

@jessesuen you asked about whether or not having malformed resources (i.e. kubectl applied) would impact the controller. I loaded the malformed resources (1x cron workflow, 1x template template, 1x cluster workflow template, 1x workflow) and I could still create workflows directly.

Problems:

* I created another valid cron workflow - this did not schedule. 
* I could not create a workflow from a valid workflow template
* I could not create a workflow from a valid cluster workflow template.

These resolved after deleting the resource - I did not need to restart the container.

# Diagnostics

```
 controller | E0805 11:42:30.478792   33530 reflector.go:153] pkg/mod/k8s.io/client-go@v0.17.5/tools/cache/reflector.go:105: Failed to list *v1alpha1.CronWorkflow: v1alpha1.CronWorkflowList.Items: []v1alpha1.CronWorkflow: v1alpha1.CronWorkflow.Spec: v1alpha1.CronWorkflowSpec.WorkflowSpec: v1alpha1.WorkflowSpec.Arguments: v1alpha1.Arguments.Parameters: []v1alpha1.Parameter: decode slice: expect [ or n, but found {, error found in #10 byte of ...|ameters":{"someParam|..., bigger context ...|spec":{"workflowSpec":{"arguments":{"parameters":{"someParam":"Hello, world!"}}}}}],"kind":"CronWork|...
 controller | E0805 11:42:30.505177   33530 reflector.go:153] pkg/mod/k8s.io/client-go@v0.17.5/tools/cache/reflector.go:105: Failed to list *v1alpha1.WorkflowTemplate: v1alpha1.WorkflowTemplateList.Items: v1alpha1.WorkflowTemplates: v1alpha1.WorkflowTemplate.Spec: v1alpha1.WorkflowTemplateSpec.WorkflowSpec: Arguments: v1alpha1.Arguments.Parameters: []v1alpha1.Parameter: decode slice: expect [ or n, but found {, error found in #10 byte of ...|ameters":{"someParam|..., bigger context ...|e81e75032120"},"spec":{"arguments":{"parameters":{"someParam":"Hello, world!"}}}}],"kind":"WorkflowT|...
```

https://github.com/kubernetes/kubernetes/issues/57705
https://github.com/argoproj/argo/issues/632

**Comment by alexec** [comment]
Options:

* Use dynamic informer to repair malformed resources. This would fight GitOps.
* Be tolerant in both listers and informers to malformed resources.

**Comment by alexec** [comment]
@jessesuen this works very nicely:

```
diff --git a/workflow/cron/controller.go b/workflow/cron/controller.go
index 57b7a35e..0f231986 100644
--- a/workflow/cron/controller.go
+++ b/workflow/cron/controller.go
@@ -12,16 +12,20 @@ import (
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/fields"
 	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/selection"
 	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/informers"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/util/workqueue"
 
+	"github.com/argoproj/argo/pkg/apis/workflow"
 	"github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
 	"github.com/argoproj/argo/pkg/client/clientset/versioned"
-	"github.com/argoproj/argo/pkg/client/informers/externalversions"
-	extv1alpha1 "github.com/argoproj/argo/pkg/client/informers/externalversions/workflow/v1alpha1"
 	"github.com/argoproj/argo/workflow/common"
 	"github.com/argoproj/argo/workflow/metrics"
 	"github.com/argoproj/argo/workflow/util"
@@ -39,7 +43,7 @@ type Controller struct {
 	wfInformer         cache.SharedIndexInformer
 	wfLister           util.WorkflowLister
 	wfQueue            workqueue.RateLimitingInterface
-	cronWfInformer     extv1alpha1.CronWorkflowInformer
+	cronWfInformer     informers.GenericInformer
 	cronWfQueue        workqueue.RateLimitingInterface
 	restConfig         *rest.Config
 	metrics            *metrics.Metrics
@@ -82,10 +86,17 @@ func (cc *Controller) Run(ctx context.Context) {
 		log.Infof("...with InstanceID: %s", cc.instanceId)
 	}
 
-	cc.cronWfInformer = externalversions.NewSharedInformerFactoryWithOptions(cc.wfClientset, cronWorkflowResyncPeriod, externalversions.WithNamespace(cc.managedNamespace),
-		externalversions.WithTweakListOptions(func(options *v1.ListOptions) {
-			cronWfInformerListOptionsFunc(options, cc.instanceId)
-		})).Argoproj().V1alpha1().CronWorkflows()
+	dc, err := dynamic.NewForConfig(cc.restConfig)
+	if err != nil {
+		log.Fatalf("failed to created dynamic interface: %w", err)
+	}
+
+	resource := dynamicinformer.NewFilteredDynamicSharedInformerFactory(dc, cronWorkflowResyncPeriod, cc.managedNamespace, func(options *v1.ListOptions) {
+		cronWfInformerListOptionsFunc(options, cc.instanceId)
+	}).
+		ForResource(schema.GroupVersionResource{Group: workflow.Group, Version: "v1alpha1", Resource: workflow.CronWorkflowPlural})
+
+	cc.cronWfInformer = resource
 	cc.addCronWorkflowInformerHandler()
 
 	cc.wfInformer = util.NewWorkflowInformer(cc.restConfig, cc.managedNamespace, cronWorkflowResyncPeriod, func(options *v1.ListOptions) {
@@ -141,9 +152,15 @@ func (cc *Controller) processNextCronItem() bool {
 		return true
 	}
 
-	cronWf, ok := obj.(*v1alpha1.CronWorkflow)
+	un, ok := obj.(*unstructured.Unstructured)
 	if !ok {
-		log.Warnf("Key '%s' in index is not a CronWorkflow", key)
+		log.Warnf("Key '%s' in index is not a Unstructured", key)
+		return true
+	}
+	cronWf := &v1alpha1.CronWorkflow{}
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(un.Object, cronWf)
+	if err != nil {
+		log.WithError(err).Error("could not convert unstructured to cron workflow")
 		return true
 	}
 
```

**Comment by alexec** [comment]
Available for testing in [v2.10.0-rc1](https://github.com/argoproj/argo/releases/tag/v2.11.0-rc1).
